### ISSUE TITLE
How to finetune Qwen-1.5/DeepSeek-1.5B parameter models

### ISSUE THREAD

**Issue by Adefioye** [allenai/open-instruct#498]
Would love to perform SFT, DPO, and RLVR on models(1.5B parameter models) not used in the study using the recipes in the codebase. Can someone please provide me with a succint guide on what to do, where do I place the models?

**Comment by hamishivi** [comment]
Hi @Adefioye ! I would start by looking at the commands in the [Tulu 3 recipe documentation](https://github.com/allenai/open-instruct/blob/main/docs/tulu3.md). You'll need to swap out the starting model with the Qwen models, and then run the training yourself (you'll need access to some GPUs).

One final thing is that you might need to make sure things are being tokenized correctly in [finetuning](https://github.com/allenai/open-instruct/blob/main/open_instruct/finetune.py), currently I think our code currently hasn't been tested on Qwen models (afaik). You can see some of the logic for handling different tokenizers [here](https://github.com/allenai/open-instruct/blob/main/open_instruct/finetune.py#L636-L668).

Let me know if you have further questions!

**Comment by hamishivi** [comment]
Hiya! I haven't tried with 1.5B but for 7B the following command works:
```bash
open_instruct/finetune.py \
    --model_name_or_path Qwen/Qwen2.5-7B \
    --tokenizer_name Qwen/Qwen2.5-7B \
    --chat_template tulu \
    --use_flash_attn \
    --use_liger_kernel True \
    --max_seq_length 16384 \
    --per_device_train_batch_size 1 \
    --gradient_accumulation_steps 2 \
    --learning_rate 5e-06 \
    --lr_scheduler_type linear \
    --warmup_ratio 0.03 \
    --weight_decay 0.0 \
    --gradient_checkpointing \
    --num_train_epochs 2 \
    --output_dir /output \
    --with_tracking \
    --report_to wandb \
    --logging_steps 1 \
    --reduce_loss sum \
    --model_revision main \
    --dataset_mixer_list allenai/tulu-3-sft-mixture 1.0 \
    --dataset_mix_dir output \
    --exp_name qwen-2-tulu-3-baseline \
    --seed 123
```

Just using `Qwen/Qwen2.5-1.5B` hopefully should just work :)

**Comment by FloatFrank** [comment]
Thank you for your reply. I used the following Linux command for dual-card training, but the loss is not decreasing. During training, it seems that tokens like <|user|> were not correctly segmented. I'm not sure where the issue lies and would appreciate your help!
```shell
#!/bin/bash
# modify the following `MACHINE_RANK`, `MAIN_PROCESS_IP`,
# `NUM_MACHINES`, `NUM_PROCESSES`, `PER_DEVICE_TRAIN_BATCH_SIZE`,
# `GRADIENT_ACCUMULATION_STEPS` according to your setup
MACHINE_RANK=0
MAIN_PROCESS_IP=localhost
NUM_MACHINES=1
NUM_PROCESSES=2
PER_DEVICE_TRAIN_BATCH_SIZE=2
GRADIENT_ACCUMULATION_STEPS=32
accelerate launch \
    --mixed_precision bf16 \
    --num_machines $NUM_MACHINES \
    --num_processes $NUM_PROCESSES \
    --machine_rank $MACHINE_RANK \
    --main_process_ip $MAIN_PROCESS_IP \
    --main_process_port 29400 \
    --use_deepspeed \
    --deepspeed_config_file configs/ds_configs/stage3_no_offloading_accelerate.conf \
    --deepspeed_multinode_launcher standard open_instruct/finetune.py \
    --model_name_or_path Qwen/Qwen2.5-1.5B \
    --tokenizer_name Qwen/Qwen2.5-1.5B \
    --use_slow_tokenizer False \
    --use_flash_attn \
    --max_seq_length 4096 \
    --preprocessing_num_workers 128 \
    --per_device_train_batch_size $PER_DEVICE_TRAIN_BATCH_SIZE \
    --gradient_accumulation_steps $GRADIENT_ACCUMULATION_STEPS \
    --learning_rate 5e-06 \
    --lr_scheduler_type linear \
    --warmup_ratio 0.03 \
    --weight_decay 0.0 \
    --num_train_epochs 2 \
    --output_dir /root/autodl-tmp/output/sft_1.5b \
    --with_tracking \
    --report_to wandb \
    --logging_steps 1 \
    --reduce_loss sum \
    --model_revision main \
    --dataset_mixer_list allenai/tulu-3-sft-mixture 1.0 \
    --checkpointing_steps epoch \
    --dataset_mix_dir output/sft_8b \
    --exp_name tulu-3-8b-sft \
    --dataset_local_cache_dir /root/autodl-tmp/cache/local_dataset_cache\
    --seed 123
```

**Comment by hamishivi** [comment]
> it seems that tokens like <|user|> were not correctly segmented

What do you mean by this? Sharing loss curves is also appreciated. We usually observe that loss only really drops at epoch boundaries, see this example wandb run replicating the Tulu 3 SFT for a reference: https://wandb.ai/ai2-llm/open_instruct_public/runs/v03nj4d2?nw=nwuserhamishivi

So your model might be learning fine - maybe it's worth checking some downstream evaluations :)

**Comment by FloatFrank** [comment]
Thank you for your reply. Since the loss was not averaged and was as high as 700, I thought it might be due to differences in the tokenizer. I will conduct tests and provide feedback.

**Comment by hamishivi** [comment]
High loss is expected with the default sum loss - this is a quirk of our setup (we used a sum loss to avoid the gradient accumulation issues that later got fixed in accelerate - https://unsloth.ai/blog/gradient has a good explainer of the issue).

**Comment by FloatFrank** [comment]
Thank you for your reply. I fully understand everything now—this is an excellent project!
